### DISCUSSION TITLE
Exposing Android ScanSettings

### DISCUSSION THREAD

**Discussion by mtrewartha** [JuulLabs/kable#158] [Ideas] [answered]
Our app makes constant use of Bluetooth and involves a lot of scanning for nearby devices. In several parts of our app, scanning needs to return results as fast as possible, which means using `ScanSettings` on Android and specifically setting the mode to the low latency/high power mode and turning reporting delay off like so:
```kotlin
ScanSettings.Builder()
    .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
    .setReportDelay(0L)
    .build()
```
It looks like the current implementation of `AndroidScanner` in this lib just sets default values for scan settings like so:
```kotlin
ScanSettings.Builder().build()
```
which makes sense.

That said, it would be really nice to be able to specify additional scan settings like we need above. I wonder if this could happen via a new variable in the `actual` `ScannerBuilder` for Android like so:
```kotlin
public actual class ScannerBuilder {
    public actual var services: List<Uuid>? = null
    public var scanSettings: ScanSettings = ScanSettings.Builder().build() // new, defaults to existing default scan settings
    ...
```
This would allow users who need to be able to customize scan settings to instantiate their scanner in their Android module but still let those that don't need that capability instantiate their scanner in common code. That feels ideal because scan settings like these aren't really multiplatform in nature (e.g. iOS doesn't even allow for controlling a scan mode or anything like it). Thoughts?

**Comment by twyatt** [comment, marked as answer]
All really good points. As an Android-only API in Kable (to allow creating a `Scanner` with Android's `ScanSettings`) sounds totally feasible.

This seems like a fairly small change, would be happy to accept a PR if you want to take that on. Otherwise I can get to it once other higher priority bugs are addressed.

**Reply by mtrewartha** [reply]
Yup, I can definitely take that on. I'll open a PR tomorrow (most likely). Thanks!
